### ISSUE TITLE
Suggestions

### ISSUE THREAD

**Issue by canatojr** [Behaviour-Analytics/moodle-block_behaviour#18]
Hello!

I'll post here some suggestions (not issues) for you to analyze and try to implement or discard if you don't like them.
Let me know in case you prefer receiving suggestions by mail instead by this moodle tracker area.

1. It would be useful if there were links from one plugin page to another. So, once on "View student behaviour" page could be links to go to "Replay clustering", "Configure resource nodes", etc ... and so on.

2.  How about having comment fields for the replay clusters as a whole? Like the student centroids and clustering centroids comments fields, could have a comment field to save infomation of each analyses like "this is the arrangement of student from group 1 build with 3 clusters".

3.  It would be better if the option of showing the students' names could be made by the teacher and not by the site administrator ... or by both with the administrator enabling or not the possibility of this feature inside his course. The same could be done with the option "grant or revoke the role of researcher to user per course" ... with users in the role of professor receiving the grant of researcher by default.

4. The "export of all data" feature could allow an option to researcher to hide or not hide the real student id.

5. The time slider is an interesting feature, but the name is not helping. Could be something more directly related with the  number of links as "link history". On the other hand, it would also be great to have another kind of slider that could permit select log periods (a log history slider).

**Comment by tedkrahn** [comment]
Thanks for the suggestions. These may take some time to implement. I have added navigation links to the pages. Also, for issue 17, I have added a feature to select students by groups. These features are in this repository, but have not yet been forwarded to the Moodle plugin directory. If you don't find any problems with the new features, I will submit to Moodle. Thanks for your help in improving the plugin!

**Comment by canatojr** [comment]
Thanks for the new features. They bring much more agility to the use of the plugin.
I tested the new version of the plugin and it is working fine. In case you find it is worth to adjust the code, in the  image you can see two details.
![behaviour_image](https://user-images.githubusercontent.com/16598643/105695142-801cfa00-5ee0-11eb-9919-3397d779917f.png)

**Comment by tedkrahn** [comment]
I have added the delete data link. I am having trouble reproducing the vertical line. It happened to me once in a course, but does not seem to be repeatable. Please let me know if you continue to see the vertical line and I will look into it some more.

**Comment by canatojr** [comment]
Yes, I still see this vertical line in all courses without built groups, but with at least one student enrolled in it and provided that at least one student has clicked on an activity / resource.

**Comment by tedkrahn** [comment]
I am still unable to reproduce the vertical line. If you open your browser's console, is there any JavaScript error message?

**Comment by canatojr** [comment]
No errors in my browser's console (image 1), but looking at the elements, I found that the vertical line is associated with the 120 px of group selection height (image 2). If I force 0 as a value (image 3), the vertical line will go away.

Looks like the code is calling this line even with no groups in the course: `<select multiple="" id="group-select" style="height: 120px;" class=""></select>`

Does this information helps in any way?
<hr>



![image1_behaviour_group_select_consolenoerror](https://user-images.githubusercontent.com/16598643/105763511-9fde0d80-5f34-11eb-8c9f-adb03fd50baa.png)
![image2_behaviour_group_select_elements_height120](https://user-images.githubusercontent.com/16598643/105763529-a7051b80-5f34-11eb-9981-a6b07886e25b.png)
![image3_behaviour_group_select_elements_height0](https://user-images.githubusercontent.com/16598643/105763540-ab313900-5f34-11eb-8630-7cdb390d03c0.png)

**Comment by tedkrahn** [comment]
Weird. I still can't reproduce this issue. The line is showing above the cluster button, which it should not be, as the button should appear on top. I looked at the HTML from my page and could not find the empty select menu anywhere. The code tests for groups and does not render the group menu when there are none (except in you case, evidently). I'm not sure, but my guess is that the groups array is not empty in your case, for some reason, which causes the code to execute and render an empty menu. But, that the menu is empty implies that there are no groups, so the array must be empty. I'm baffled.

**Comment by tedkrahn** [comment]
Is it possible for you to query the groups table? What is in there?

**Comment by canatojr** [comment]
Yes it is really weird! There is nothing in groups table!

To make sure that the problem was not related to any conflict with another plugin or customization of the site, I installed moodle 3.10+ in a new empty database and the problem persists.

So, let me resume the sequence that I have done:

- Installed Moodle version: 3.10+ (Build: 20201127) in an empty database
- Installed Behaviour version: 0.7.5 (2021012501) - the unique aditional plugin installed
- created one course
- created 3 users
- enroled that 3 users in the course as students
- inserted 2 page resources
- logged in with those 3 students and did some browsing between those pages resources and announcements forum
- enroled me (the admin) as a teacher in the course and granted me researcher status (yes, I now that as admin don't need that)
- inserted behaviour block in the course
- clicked on view students behaviour link 
- and ... the vertical line is there ... sure does not have any group entry in database.
- the result is the same using Chrome, Edge and Firefox (except that with Firefox the vertical line is thicker).

At least is a minor problem, just stetical!

**Comment by tedkrahn** [comment]
So bizarre. I just can't reproduce this issue. And you are running the latest version (20210012501). Is it possible for you to insert a "console.log(groups);" statement at line 2100 of /blocks/behaviour/amd/src/behaviour-analytics.js? Site Administration -> Appearance -> Cache JavaScript will need to be off for the changes to take effect. What is the browser console output?

**Comment by canatojr** [comment]
Desabling the JavaScript Cache the vertical line goes away (with or without the insert indicated in line 2100).
And when inserting the line indicated in behavior-analytics.js, the output from the browser console does not show errors, but information about line 2100, as shown in the attached image.
Let me know if you need more information.
<hr>


![image4_behaviour_consoleline2100](https://user-images.githubusercontent.com/16598643/105977760-84781d00-6070-11eb-8f93-73a5cc8a8f5e.png)

**Comment by tedkrahn** [comment]
I believe I have finally figured it out. As it turns out, I have trouble with simple programming tasks sometimes. It looks like I forgot to minify the latest changes in the JavaScript. When caching is on, the bug is there because the minified code is running. With the caching off, the human readable code is running. These were slightly different. I have corrected the problem, so the last commit should take care of this issue for. Please let me know. Thanks again for all your help.

**Comment by canatojr** [comment]
All done!
Congratulations!
Sometimes, what seems to be the easiest is the hardest part!

**Comment by tedkrahn** [comment]
I have added suggestion 2, ability to name/comment on each cluster analysis. When a cluster analysis is made or replayed, it is now possible to name it or add a comment. The text box and save button are added to the navigation menu. I also fixed the problem of the clustering and replay log buttons being inside the log panel. Please let me know if these cause you and problems. Thanks again!

**Comment by canatojr** [comment]
Everything seems to be working fine!
Thanks!
